File: csrc/pytorch/mega_moe/mega_moe.cpp
// Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
//
// See LICENSE for license information.

#include <ATen/cuda/CUDAContext.h>
#include <torch/extension.h>

#include "primus_turbo/mega_moe.h"

#include "../extensions.h"

namespace primus_turbo::pytorch {

using primus_turbo::mega_moe::get_mega_moe_config;
using primus_turbo::mega_moe::kScaleGroupK;
using primus_turbo::mega_moe::launch_fp8_fp4_mega_moe;
using primus_turbo::mega_moe::MegaMoEArgs;
using primus_turbo::mega_moe::MegaMoEBufferLayout;
using primus_turbo::mega_moe::MegaMoEConfig;
namespace mm = primus_turbo::mega_moe;

// Mirrors DeepGEMM's ``deep_gemm::mega::get_token_alignment_for_mega_moe``.
int64_t get_token_alignment_for_mega_moe() {
    return primus_turbo::mega_moe::kTokenAlignment;
}

// Returns the symmetric-buffer byte size plus the offsets of each
// logical region, packed as a 1-D int64 tensor laid out as:
//   [ total_bytes,
//     num_max_pool_tokens, num_padded_sf_pool_tokens,
//     workspace_offset,
//     input_x_offset, input_x_sf_offset,
//     input_topk_idx_offset, input_topk_weights_offset,
//     l1_pool_x_offset, l1_pool_x_sf_offset, l1_pool_weights_offset,
//     l2_pool_x_offset, l2_pool_x_sf_offset,
//     combine_buffer_offset ]
//
// The Python side then slices these offsets into actual tensor views
// over the symmetric memory buffer.  Mirrors DeepGEMM's
// ``deep_gemm::mega::get_symm_buffer_size_for_mega_moe``.  DG returns
// ``(num_bytes, slice_input_buffers_callback)``; we return the raw
// offsets so the Python layer (registered through ``torch.library``)
// can slice them without a second C++ round-trip.
at::Tensor get_symm_buffer_size_for_mega_moe(const int64_t num_ranks, const int64_t num_experts,
                                             const int64_t num_max_tokens_per_rank,
                                             const int64_t num_topk, const int64_t hidden,
                                             const int64_t intermediate_hidden,
                                             const bool    use_fp8_dispatch) {
    const auto layout = mm::get_symm_buffer_size_for_mega_moe(
        static_cast<int>(num_ranks), static_cast<int>(num_experts),
        static_cast<int>(num_max_tokens_per_rank), static_cast<int>(num_topk),
        static_cast<int>(hidden), static_cast<int>(intermediate_hidden), use_fp8_dispatch);

    at::Tensor out = at::empty({14}, at::dtype(at::kLong).device(at::kCPU));
    auto       p   = out.data_ptr<int64_t>();
    p[0]           = layout.total_bytes;
    p[1]           = layout.num_max_pool_tokens;
    p[2]           = layout.num_padded_sf_pool_tokens;
    p[3]           = layout.workspace_offset;
    p[4]           = layout.input_x_offset;
    p[5]           = layout.input_x_sf_offset;
    p[6]           = layout.input_topk_idx_offset;
    p[7]           = layout.input_topk_weights_offset;
    p[8]           = layout.l1_pool_x_offset;
    p[9]           = layout.l1_pool_x_sf_offset;
    p[10]          = layout.l1_pool_weights_offset;
    p[11]          = layout.l2_pool_x_offset;
    p[12]          = layout.l2_pool_x_sf_offset;
    p[13]          = layout.combine_buffer_offset;
    return out;
}

// Fused mega-MoE FP8 activations × FP4 weights entry.  Currently a thin
// host-side validation shell that forwards to ``launch_fp8_fp4_mega_moe``
// (which is itself a no-op stub for unsupported AOT shapes).  All input
// tensors are expected to live in symmetric memory or device memory.
// Mirrors DeepGEMM's ``deep_gemm::mega::fp8_fp4_mega_moe``.
void fp8_fp4_mega_moe(at::Tensor y, at::Tensor l1_weights, at::Tensor l1_weights_sf,
                      at::Tensor l2_weights, at::Tensor l2_weights_sf,
                      c10::optional<at::Tensor> cumulative_local_expert_recv_stats,
                      at::Tensor sym_buffer, const std::vector<int64_t> &sym_buffer_ptrs,
                      const int64_t rank_idx, const int64_t num_max_tokens_per_rank,
                      const int64_t num_experts, const int64_t num_topk, const int64_t num_tokens,
                      const int64_t hidden, const int64_t intermediate_hidden,
                      const std::vector<int64_t> &recipe, const std::string &activation,
                      const double activation_clamp, const bool fast_math) {
    TORCH_CHECK(activation == "swiglu", "mega_moe: only swiglu activation is currently supported");
    TORCH_CHECK(recipe.size() == 3, "mega_moe: recipe must be a length-3 tuple (M, N, K)");
    TORCH_CHECK(recipe[0] == 1 && recipe[1] == 1 && recipe[2] == kScaleGroupK,
                "mega_moe: only recipe=(1, 1, ", kScaleGroupK, ") is currently supported, got (",
                recipe[0], ", ", recipe[1], ", ", recipe[2], ")");
    TORCH_CHECK(activation_clamp >= 0, "mega_moe: activation_clamp must be non-negative, got ",
                activation_clamp);
    TORCH_CHECK(y.is_cuda(), "mega_moe: output tensor must live on GPU");
    TORCH_CHECK(sym_buffer.is_cuda(), "mega_moe: symmetric buffer must live on GPU");
    TORCH_CHECK(!sym_buffer_ptrs.empty(), "mega_moe: sym_buffer_ptrs must be non-empty");
    TORCH_CHECK(num_max_tokens_per_rank % primus_turbo::mega_moe::kTokenAlignment == 0,
                "mega_moe: num_max_tokens_per_rank (", num_max_tokens_per_rank,
                ") must be a multiple of kTokenAlignment (",
                primus_turbo::mega_moe::kTokenAlignment, ")");

    const int num_ranks = static_cast<int>(sym_buffer_ptrs.size());
    TORCH_CHECK(num_experts % num_ranks == 0,
                "mega_moe: num_experts must be divisible by num_ranks");
    const int num_experts_per_rank = static_cast<int>(num_experts) / num_ranks;

    // Shape validation: mirror DG's ``check_grouped_ab_fp8_fp4`` semantics.
    TORCH_CHECK(l1_weights.dim() == 3 && l2_weights.dim() == 3,
                "mega_moe: l1/l2 weights must be 3-D [num_experts_per_rank, N, K_packed]");
    TORCH_CHECK(l1_weights.size(0) == num_experts_per_rank &&
                    l2_weights.size(0) == num_experts_per_rank,
                "mega_moe: l1/l2 weights group dim must equal num_experts_per_rank (",
                num_experts_per_rank, ")");
    TORCH_CHECK(l1_weights.size(1) == 2 * intermediate_hidden,
                "mega_moe: l1_weights N must equal 2*intermediate_hidden (",
                2 * intermediate_hidden, "), got ", l1_weights.size(1));
    TORCH_CHECK(l2_weights.size(1) == hidden, "mega_moe: l2_weights N must equal hidden (", hidden,
                "), got ", l2_weights.size(1));
    TORCH_CHECK(l1_weights.is_contiguous() && l2_weights.is_contiguous(),
                "mega_moe: l1/l2 weights must be contiguous");
    TORCH_CHECK(y.dim() == 2 && y.size(1) == hidden, "mega_moe: y must be 2D with hidden dim ",
                hidden, ", got shape ", y.sizes());
    TORCH_CHECK(num_tokens <= num_max_tokens_per_rank, "mega_moe: num_tokens (", num_tokens,
                ") exceeds num_max_tokens_per_rank (", num_max_tokens_per_rank, ")");

    if (cumulative_local_expert_recv_stats.has_value()) {
        const auto &stats = cumulative_local_expert_recv_stats.value();
        TORCH_CHECK(stats.scalar_type() == at::kInt,
                    "mega_moe: cumulative_local_expert_recv_stats must be int32");
        TORCH_CHECK(stats.numel() == num_experts_per_rank,
                    "mega_moe: cumulative_local_expert_recv_stats must have ", num_experts_per_rank,
                    " elements");
        TORCH_CHECK(stats.is_contiguous(),
                    "mega_moe: cumulative_local_expert_recv_stats must be contiguous");
    }

    const auto layout = mm::get_symm_buffer_size_for_mega_moe(
        num_ranks, static_cast<int>(num_experts), static_cast<int>(num_max_tokens_per_rank),
        static_cast<int>(num_topk), static_cast<int>(hidden), static_cast<int>(intermediate_hidden),
        /*use_fp8_dispatch=*/true);
    TORCH_CHECK(sym_buffer.nbytes() >= static_cast<size_t>(layout.total_bytes),
                "mega_moe: symmetric buffer too small for requested config");

    auto config = get_mega_moe_config(
        num_ranks, static_cast<int>(num_experts), num_experts_per_rank,
        static_cast<int>(num_max_tokens_per_rank), static_cast<int>(num_tokens),
        static_cast<int>(num_topk), static_cast<int>(hidden), static_cast<int>(intermediate_hidden),
        layout.num_padded_sf_pool_tokens);
    config.num_max_pool_tokens = layout.num_max_pool_tokens;

    MegaMoEArgs args;
    args.y_ptr             = y.data_ptr();
    args.l1_weights_ptr    = l1_weights.data_ptr();
    args.l1_weights_sf_ptr = l1_weights_sf.data_ptr();
    args.l2_weights_ptr    = l2_weights.data_ptr();
    args.l2_weights_sf_ptr = l2_weights_sf.data_ptr();
    args.cumulative_local_expert_recv_stats =
        cumulative_local_expert_recv_stats.has_value()
            ? cumulative_local_expert_recv_stats->data_ptr<int>()
            : nullptr;
    args.sym_buffer_ptrs         = sym_buffer_ptrs.data();
    args.num_ranks               = num_ranks;
    args.rank_idx                = static_cast<int>(rank_idx);
    args.layout                  = layout;
    args.num_tokens              = static_cast<int>(num_tokens);
    args.num_max_tokens_per_rank = static_cast<int>(num_max_tokens_per_rank);
    args.num_experts             = static_cast<int>(num_experts);
    args.num_experts_per_rank    = num_experts_per_rank;
    args.num_topk                = static_cast<int>(num_topk);
    args.hidden                  = static_cast<int>(hidden);
    args.intermediate_hidden     = static_cast<int>(intermediate_hidden);
    args.recipe_m                = static_cast<int>(recipe[0]);
    args.recipe_n                = static_cast<int>(recipe[1]);
    args.recipe_k                = static_cast<int>(recipe[2]);
    args.activation_clamp        = static_cast<float>(activation_clamp);
    args.fast_math               = fast_math;
    args.config                  = config;
    args.stream                  = at::cuda::getCurrentCUDAStream();

    launch_fp8_fp4_mega_moe(args);
}

} // namespace primus_turbo::pytorch

File: primus_turbo/pytorch/ops/mega_moe.py
###############################################################################
# Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
#
# See LICENSE for license information.
###############################################################################
"""Public API for the fused Mega MoE FFN kernel.

Mirrors DeepGEMM's ``deep_gemm.mega`` API surface 1:1 so existing call
sites in DeepSeek-style MoE pipelines can switch backends with minimal
diff:

* :class:`SymmBuffer`                    — symmetric memory + tensor views
* :func:`get_symm_buffer_for_mega_moe`   — factory matching the DG name
* :func:`transform_weights_for_mega_moe` — weight pre-processing
* :func:`fp8_fp4_mega_moe`               — fused dispatch + GG1 + SwiGLU + GG2 + combine

The underlying GPU kernel is a stub in this revision; the Python layer
already wires up symmetric memory rendezvous, buffer slicing and shape
validation so that downstream integrations can be written against the
final API.
"""

from __future__ import annotations

from typing import Optional, Tuple

import torch
import torch.distributed as dist

from primus_turbo.pytorch.core.symm_mem import SymmetricMemory
from primus_turbo.pytorch.kernels.mega_moe import (
    SymmBufferLayout,
    fp8_fp4_mega_moe_impl,
    get_symm_buffer_size_for_mega_moe,
    get_token_alignment_for_mega_moe,
    transform_l1_weights_for_mega_moe,
    transform_l2_weights_for_mega_moe,
)

__all__ = [
    "SymmBuffer",
    "fp8_fp4_mega_moe",
    "get_symm_buffer_for_mega_moe",
    "transform_weights_for_mega_moe",
]


def _align_up(value: int, multiple: int) -> int:
    return ((value + multiple - 1) // multiple) * multiple


# ---------------------------------------------------------------------------
# Symmetric buffer wrapper
# ---------------------------------------------------------------------------


class SymmBuffer:
    """Container holding the symmetric memory allocation + tensor views.

    Equivalent of DeepGEMM's ``deep_gemm.mega.SymmBuffer``.  Each rank
    in ``group`` calls into this class with the same arguments; the
    constructor rendezvous'es IPC handles internally and exposes
    per-region tensor views that the caller can copy inputs into
    before invoking :func:`fp8_fp4_mega_moe`.

    The exposed views mirror DG's ``slice_input_buffers`` exactly:
    ``(x, x_sf, topk_idx, topk_weights, l1_acts, l1_acts_sf, l2_acts,
    l2_acts_sf)``.  Internal regions (per-expert weight columns,
    BF16 combine buffer) are not exposed — the kernel addresses them
    directly through ``sym_buffer_ptrs``.
    """

    def __init__(
        self,
        group: dist.ProcessGroup,
        *,
        num_experts: int,
        num_max_tokens_per_rank: int,
        num_topk: int,
        hidden: int,
        intermediate_hidden: int,
        use_fp8_dispatch: bool = True,
        activation: str = "swiglu",
    ) -> None:
        if activation != "swiglu":
            raise NotImplementedError(f"SymmBuffer only supports activation='swiglu', got {activation!r}")

        self.group = group
        self.num_experts = int(num_experts)
        self.num_max_tokens_per_rank = int(num_max_tokens_per_rank)
        self.num_topk = int(num_topk)
        self.hidden = int(hidden)
        self.intermediate_hidden = int(intermediate_hidden)
        self.use_fp8_dispatch = bool(use_fp8_dispatch)
        self.activation = activation

        # Per-rank token count must be aligned.
        self.num_max_tokens_per_rank = _align_up(
            self.num_max_tokens_per_rank, get_token_alignment_for_mega_moe()
        )

        # Compute the layout (offsets + total bytes).
        self.layout: SymmBufferLayout = get_symm_buffer_size_for_mega_moe(
            num_ranks=group.size(),
            num_experts=self.num_experts,
            num_max_tokens_per_rank=self.num_max_tokens_per_rank,
            num_topk=self.num_topk,
            hidden=self.hidden,
            intermediate_hidden=self.intermediate_hidden,
            use_fp8_dispatch=self.use_fp8_dispatch,
        )

        # Allocate symmetric memory.  ``SymmetricMemory`` handles IPC
        # rendezvous and exposes per-rank device pointers.
        self.symm_mem = SymmetricMemory(group, alloc_size=int(self.layout.total_bytes))
        # ``buffer`` is a flat int8 view over this rank's local slice.
        self.buffer: torch.Tensor = self.symm_mem.get_buffer(
            rank=self.symm_mem.rank,
            sizes=[int(self.layout.total_bytes)],
            dtype=torch.int8,
        )

        # Create per-region tensor views (mirrors DG's 8-tuple).
        (
            self.x,
            self.x_sf,
            self.topk_idx,
            self.topk_weights,
            self.l1_acts,
            self.l1_acts_sf,
            self.l2_acts,
            self.l2_acts_sf,
        ) = self._slice_views()

        # Initialise to zero and barrier so peers see a clean buffer.
        self.buffer.zero_()
        self.group.barrier()
        torch.cuda.synchronize()

    # ------------------------------------------------------------------
    #  Convenience properties
    # ------------------------------------------------------------------

    @property
    def num_ranks(self) -> int:
        return self.symm_mem.world_size

    @property
    def rank(self) -> int:
        return self.symm_mem.rank

    @property
    def buffer_ptrs(self) -> list[int]:
        """Per-rank base device pointers used by the launcher."""
        return list(self.symm_mem.buffer_ptrs)

    # ------------------------------------------------------------------
    #  Internal helpers
    # ------------------------------------------------------------------

    def _view_at(
        self,
        offset_bytes: int,
        shape: Tuple[int, ...],
        dtype: torch.dtype,
    ) -> torch.Tensor:
        """Return a typed view over the local symmetric buffer."""
        if offset_bytes % dtype.itemsize != 0:
            raise ValueError(
                f"SymmBuffer: offset {offset_bytes} is not aligned to "
                f"dtype {dtype} (itemsize={dtype.itemsize})"
            )
        storage_offset = offset_bytes // dtype.itemsize
        return self.symm_mem.get_buffer(
            rank=self.rank, sizes=list(shape), dtype=dtype, storage_offset=storage_offset
        )

    def _slice_views(
        self,
    ) -> Tuple[
        torch.Tensor,  # x          FP8   [N, H]
        torch.Tensor,  # x_sf       int   [N, H/128]
        torch.Tensor,  # topk_idx   int64 [N, K]
        torch.Tensor,  # topk_w     f32   [N, K]
        torch.Tensor,  # l1_acts    FP8   [P, H]
        torch.Tensor,  # l1_acts_sf int   [P_sf, H/128]
        torch.Tensor,  # l2_acts    FP8   [P, I]
        torch.Tensor,  # l2_acts_sf int   [P_sf, I/128]
    ]:
        layout = self.layout

        x_dtype = torch.float8_e4m3fnuz if self.use_fp8_dispatch else torch.bfloat16

        x = self._view_at(
            layout.input_x_offset,
            shape=(self.num_max_tokens_per_rank, self.hidden),
            dtype=x_dtype,
        )
        x_sf = self._view_at(
            layout.input_x_sf_offset,
            shape=(self.num_max_tokens_per_rank, self.hidden // 128),
            dtype=torch.int32,
        )
        topk_idx = self._view_at(
            layout.input_topk_idx_offset,
            shape=(self.num_max_tokens_per_rank, self.num_topk),
            dtype=torch.int64,
        )
        topk_weights = self._view_at(
            layout.input_topk_weights_offset,
            shape=(self.num_max_tokens_per_rank, self.num_topk),
            dtype=torch.float32,
        )

        l1_acts = self._view_at(
            layout.l1_pool_x_offset,
            shape=(layout.num_max_pool_tokens, self.hidden),
            dtype=x_dtype,
        )
        l1_acts_sf = self._view_at(
            layout.l1_pool_x_sf_offset,
            shape=(layout.num_padded_sf_pool_tokens, self.hidden // 128),
            dtype=torch.int32,
        )
        l2_acts = self._view_at(
            layout.l2_pool_x_offset,
            shape=(layout.num_max_pool_tokens, self.intermediate_hidden),
            dtype=x_dtype,
        )
        l2_acts_sf = self._view_at(
            layout.l2_pool_x_sf_offset,
            shape=(layout.num_padded_sf_pool_tokens, self.intermediate_hidden // 128),
            dtype=torch.int32,
        )
        return (
            x,
            x_sf,
            topk_idx,
            topk_weights,
            l1_acts,
            l1_acts_sf,
            l2_acts,
            l2_acts_sf,
        )

    def destroy(self) -> None:
        """Release the symmetric memory + IPC handles."""
        # Drop tensor refs so torch can release backing storage.
        for attr in (
            "x",
            "x_sf",
            "topk_idx",
            "topk_weights",
            "l1_acts",
            "l1_acts_sf",
            "l2_acts",
            "l2_acts_sf",
            "buffer",
        ):
            setattr(self, attr, None)
        if self.symm_mem is not None:
            self.symm_mem.destroy()
            self.symm_mem = None


# ---------------------------------------------------------------------------
# Factories / public ops
# ---------------------------------------------------------------------------


def get_symm_buffer_for_mega_moe(
    group: dist.ProcessGroup,
    *,
    num_experts: int,
    num_max_tokens_per_rank: int,
    num_topk: int,
    hidden: int,
    intermediate_hidden: int,
    use_fp8_dispatch: bool = True,
    activation: str = "swiglu",
) -> SymmBuffer:
    """Allocate the symmetric buffer used by :func:`fp8_fp4_mega_moe`.

    Mirrors DG's ``deep_gemm.mega.get_symm_buffer_for_mega_moe``.
    """

    return SymmBuffer(
        group,
        num_experts=num_experts,
        num_max_tokens_per_rank=num_max_tokens_per_rank,
        num_topk=num_topk,
        hidden=hidden,
        intermediate_hidden=intermediate_hidden,
        use_fp8_dispatch=use_fp8_dispatch,
        activation=activation,
    )


def transform_weights_for_mega_moe(
    l1_weights: Tuple[torch.Tensor, torch.Tensor],
    l2_weights: Tuple[torch.Tensor, torch.Tensor],
) -> Tuple[Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor, torch.Tensor]]:
    """Transform (weight, scaling-factor) pairs into the kernel-native layout.

    * L1: interleave gate / up along N then permute the SF tensor.
    * L2: permute the SF tensor only.

    Match DeepGEMM's ``transform_weights_for_mega_moe`` semantics so
    pre-trained weight checkpoints can be shared.
    """

    return (
        transform_l1_weights_for_mega_moe(l1_weights),
        transform_l2_weights_for_mega_moe(l2_weights),
    )


def fp8_fp4_mega_moe(
    y: torch.Tensor,
    l1_weights: Tuple[torch.Tensor, torch.Tensor],
    l2_weights: Tuple[torch.Tensor, torch.Tensor],
    sym_buffer: SymmBuffer,
    *,
    cumulative_local_expert_recv_stats: Optional[torch.Tensor] = None,
    recipe: Tuple[int, int, int] = (1, 1, 32),
    activation: str = "swiglu",
    activation_clamp: Optional[float] = None,
    fast_math: bool = True,
) -> None:
    """Run the fused mega-MoE FFN (dispatch + GG1 + SwiGLU + GG2 + combine).

    FP8 activations × FP4 weights variant.  Mirrors DG's
    ``deep_gemm.mega.fp8_fp4_mega_moe`` API surface 1:1.

    Args:
        y: BF16 output tensor of shape ``[num_tokens, hidden]``.  This
            rank's combined per-token output is written in-place.
        l1_weights: ``(weight, sf)`` pair for the up-projection.  Must
            already be in the kernel-native layout (use
            :func:`transform_weights_for_mega_moe`).
        l2_weights: ``(weight, sf)`` pair for the down-projection.
        sym_buffer: A :class:`SymmBuffer` previously allocated by
            :func:`get_symm_buffer_for_mega_moe`.  The caller is
            expected to copy ``x``, ``x_sf``, ``topk_idx`` and
            ``topk_weights`` into the buffer views before invoking.
        cumulative_local_expert_recv_stats: Optional int32 tensor of
            shape ``[num_experts_per_rank]`` accumulating the number of
            received tokens per local expert across calls.
        recipe: SF granularity ``(M, N, K)``.  Only ``(1, 1, 32)`` is
            currently supported.
        activation: Activation function name.  Only ``swiglu`` is
            currently supported.
        activation_clamp: Optional clamp value for SwiGLU stability.
            ``None`` disables clamping.
        fast_math: Enable fast-math approximations inside the activation.
    """

    if recipe != (1, 1, 32):
        raise NotImplementedError(f"fp8_fp4_mega_moe currently supports recipe=(1, 1, 32), got {recipe}")
    if y.dim() != 2:
        raise ValueError(f"fp8_fp4_mega_moe: y must be 2D, got shape {tuple(y.shape)}")
    if y.size(1) != sym_buffer.hidden:
        raise ValueError(
            f"fp8_fp4_mega_moe: y hidden dim ({y.size(1)}) does not match buffer hidden "
            f"({sym_buffer.hidden})"
        )

    num_tokens = int(y.size(0))
    if num_tokens > sym_buffer.num_max_tokens_per_rank:
        raise ValueError(
            f"fp8_fp4_mega_moe: num_tokens ({num_tokens}) exceeds "
            f"num_max_tokens_per_rank ({sym_buffer.num_max_tokens_per_rank})"
        )

    fp8_fp4_mega_moe_impl(
        y,
        l1_weights,
        l2_weights,
        sym_buffer=sym_buffer.buffer,
        sym_buffer_ptrs=sym_buffer.buffer_ptrs,
        rank_idx=sym_buffer.rank,
        num_max_tokens_per_rank=sym_buffer.num_max_tokens_per_rank,
        num_experts=sym_buffer.num_experts,
        num_topk=sym_buffer.num_topk,
        num_tokens=num_tokens,
        hidden=sym_buffer.hidden,
        intermediate_hidden=sym_buffer.intermediate_hidden,
        cumulative_local_expert_recv_stats=cumulative_local_expert_recv_stats,
        recipe=recipe,
        activation=activation,
        activation_clamp=activation_clamp,
        fast_math=fast_math,
    )
